articles/ros2_security_contexts.md

```diff
@@ -0,0 +1,315 @@
+This is reminiscent of the pattern used earlier Keymint [1].
+
+
+```
+$ tree keystore/
+keystore
+├── contexts
+│   └── ...
+│       └── ...
+├── private
+│   ├── ca.csr.pem
+│   └── ca.key.pem
+└── public
+    ├── ca.cert.pem
+    ├── identity_ca.cert.pem -> ca.cert.pem
+    └── permissions_ca.cert.pem -> ca.cert.pem
+```
+
+
+### ``public``
+
+The ``public`` directory contains anything permissible as public, such as public certificates for the identity or permissions certificate authorities.
+As such, this can be given read access to all executables.
+Note that in the default case, both the `identity_ca` and `permissions_ca` points to the same CA certificate.
+
+### ``private``
+
+The ``private`` directory contains anything permissable as private, such as private key material for aforementioned certificate authorities.
+This directory should be redacted before deploying the keystore onto the target device/robot.
+
+### ``contexts``
+
+The ``contexts`` directory contains the security artifacts associated with individual contexts, and thus node directories are no longer relevant.
+Similar to node directories however, the `contexts` folder may still recursively nest sub-paths for organizing separate contexts.
+The `ROS_SECURITY_ROOT_DIRECTORY` environment variable should by convention point to this directory.
+
+
+## Runtime
+
+Given the normative case where a context within a policy may be specific to a single node/container process, the namespace the node is remapped to will inevitably affect the required security permissions necessary within the context.
+To highlight this interdependency, and to help avoid context path collisions, a hierarchy borrowing namespaces is appropriate. 
+By convention, ros2launch could be used to prefix relative context paths for single process node or containers using the namespace in scope, to enable a convention of composable launch files with adjustable and parameterized context paths.
+Given the runtime command argument for specifying the fully qualified context path, ros2launch would accordingly resolve relative context paths for executables, as defined by launch attributes.
+
+### Unqualified context path
+
+For single process nodes with unqualified context paths, the context directory will subsequently default to the root level context.
+
+``` xml
+<launch>
+  <node pkg="demo_nodes_cpp" exec="talker"/>
+  <node pkg="demo_nodes_cpp" exec="listener"/>
+</launch>
+```
+
+```
+$ tree contexts/
+contexts/
+├── cert.pem
+├── governance.p7s
+├── identity_ca.cert.pem -> ../public/identity_ca.cert.pem
+├── key.pem
+├── permissions_ca.cert.pem -> ../public/permissions_ca.cert.pem
+└── permissions.p7s
+```
+
+### Pushed unqualified context path
+
+For single process nodes with unqualified context paths pushed by a namespace, the context directory will subsequently be pushed to the relative sub-folder.
```

Comment:
Same here, is this still something we consider?

Reply:
https://github.com/ros2/design/pull/274#discussion_r403319180

Reply:
Relegated to future work: 3e148b2